### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by lschaefer-sugarcrm**
@glevine-sugarcrm Are you able to review the unit tests?  I think I mostly stuck with what you suggested. 

@mmarum-sugarcrm This PR is finally ready for review!  If you're able to try out running the tests locally as well as in Jenkins, it might be really helpful.  I think I've documented all of the little pieces you need to install and configure, but it's possible I missed something along the way.  I've been opening Issues related to this work that can be finished separate from this PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -174,16 +174,39 @@ You can then click Details to open the build results in Travis CI.
 
 ### About the build
 
-The build is configured in [.travis.yml](.travis.yml). Currently, the build has two stages:
-- test
+The build has four Environment Variables that have been configured in the project settings in Travis CI:
+- DEV_COMMUNITY_USERNAME: The username for an account that has access to the 
+[SugarCRM Developer Builds Space](https://community.sugarcrm.com/community/developer/developer-builds)
+- DEV_COMMUNITY_PASSWORD: The password associated with the above account
+- GIT_HUB_USERNAME: The username for a GitHub account that has access to https://github.com/sugarcrm/unit-tests
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply README.md]: I actually waffled over this.  I debated if GITHUB would make you read the th together like the th in thesaurus.   Obviously, I overthought this.  Updating.

**Comment by lschaefer-sugarcrm** [comment]
@mmarum-sugarcrm I've made the updates.  Can you take another look?  The build is passing in both Jenkins and Travis.  I'm still debugging why the Prof M MLP isn't installing cleanly locally on my machine.

File: package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
+<?php
+
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php]: You should add a namespace for this test.

`namespace Sugarcrm\SugarcrmTestsUnit\modules\Schedulers\Ext\ScheduledTasks;`

File: package/src/custom/tests/School/unit-php/gradebook_fake/RecordManagerTest.php [hunk 3]
```diff
@@ -7,11 +7,13 @@
 /**
  * Class RecordManagerTest
  * Tests the RecordManager for the GradebookFake app
+ * @coversDefaultClass \RecordManager
  */
 class RecordManagerTest extends TestCase
 {
     /**
      * Check that when valid params are sent to createStudentRecord, true is returned
+     * @covers createStudentRecord
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/gradebook_fake/RecordManagerTest.php]: Need to have `::` before the method name.

File: package/src/custom/tests/School/unit-php/gradebook_fake/RecordManagerTest.php [hunk 4]
```diff
@@ -21,6 +23,7 @@ public function testRecordManagerValidParams(){
     /**
      * Check that when the email address forceerror@example.com is used as a param for createStudentRecord, an
      * exception is thrown
+     * @covers createStudentRecord
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/gradebook_fake/RecordManagerTest.php]: Need to have `::` before the method name.

File: package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+use Sugarcrm\SugarcrmTestsUnit\TestMockHelper;
+use Sugarcrm\Sugarcrm\Util\Uuid;
+
+require_once 'include/SugarQueue/SugarJobQueue.php';
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php]: Another case of including `SugarJobQueue` that I really don't think should be necessary.

File: package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * @coversDefaultClass \Students_Gradebook
+ */
+class Student_GradebookTest extends \PHPUnit_Framework_TestCase
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php]: Another case where we switched to `\PHPUnit\Framework\TestCase` recently and you might need to update.

File: package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    private $sg;
+
+    /**
+     * @var a partial mock of a Student (Contact) with an id, first name, last name, and email
+     */
+    private $student;
+
+    protected function setUp()
+    {
+        \SugarAutoLoader::load('../../../modules/Contacts/Students_Gradebook.php');
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php]: Another one I would just require directly above the class. And you might not need the `../../../`.

File: package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        $this->assertSame($GLOBALS['current_user']->id, $this->job->assigned_user_id);
+    }
+
+    /**
+     * @covers ::addStudentToGradebook
+     */
+    public function testAddStudentToGradebook_DoesNotCreateTheJob_NotAfterSaveEvent()
+    {
+        $this->sg->addStudentToGradebook($this->student, 'after_relationship_add', ['isUpdate' => false]);
+        $this->queue->expects($this->never())->method('submitJob');
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php]: You need to set your expectation before you call `addStudentToGradebook`.

File: package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * @covers ::addStudentToGradebook
+     */
+    public function testAddStudentToGradebook_DoesNotCreateTheJob_IsUpdateEvent()
+    {
+        $this->sg->addStudentToGradebook($this->student, 'after_save', ['isUpdate' => true]);
+        $this->queue->expects($this->never())->method('submitJob');
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/modules/Contacts/Student_GradebookTest.php]: You need to set your expectation before you call `addStudentToGradebook`.

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    <php>
+        <ini name="intl.default_locale" value="en"/>
+        <ini name="intl.error_level" value="0"/>
+        <ini name="memory_limit" value="-1"/>
+    </php>
+    <testsuites>
+        <testsuite name="all">
+            <directory>./</directory>
+            <directory>modules/</directory>
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/phpunit.xml.dist]: Doesn't `./` includes `modules/`? It should be recursive. If you got rid of this second line, then would it still run everything?

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        </testsuite>
+    </testsuites>
+    <filter>
+        <whitelist processUncoveredFilesFromWhitelist="false">
+            <directory suffix=".php">../../data/</directory>
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/phpunit.xml.dist]: None of these directories are from your package. You only care about code coverage for the code you write, so this should probably be a list of directories or files that are installed by your package. If you tell me the paths of all of the files that are installed, then I can help you figure out what entries need to be here.

https://phpunit.de/manual/current/en/appendixes.configuration.html#appendixes.configuration.whitelisting-files

File: scripts/RunProfMPHPUnitTests.sh [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+# This script runs the Sugar PHPUnit tests. It assumes the Sugar Docker stack has already been started and that
```
Comment (glevine) [inline scripts/RunProfMPHPUnitTests.sh]: This runs "your" tests, not the Sugar tests, right?

**Comment by glevine** [comment]
@lschaefer-sugarcrm - You now have all of your custom code tested, assuming it all exists in the files...

```
src/custom/modules/Contacts/Students_Gradebook.php
src/custom/gradebook_fake/RecordManager.php
src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php
```

... as the rest looks like metadata changes. You likely have 100% code coverage and have tested all of your edge cases. So now you could write a single integration test that tests a complete happy path for the entire flow fro saving a Contact to doing whatever RecordManager is going to do (If it never really does anything, then verifying that the job was completed successfully is probably your best bet for validating the path.). With that you'll have a fairly comprehensive testing story that includes tests at various layers within the testing pyramid (e.g., most of your tests are true unit tests that help you isolate issues based on very specific conditions and a very small number of integration tests -- in this case one -- that make sure your code will produce the correct result when integrated into a system that behaves according to the assumptions upon which your code was designed).

You'll run into some problems, however. The biggest one is that we don't currently have a way of testing asynchronous flows, like those that execute using the queue. You would end up having to write a test that executes the queue and then checks the status of your job, over and over until completion. That would not be ideal as it means your test would have unnecessary complexity and the runtime could be excessive because you don't know how long it will take to run.

I think you should file a story on the AT board for creating an API that would enable executing the job queue to run a specific job or a specific set of jobs, whether it be by name or ID or something else. Then an integration test could be written that does some work, creates a job, executes only the job or jobs that matter for that test, and assert the results.

File: package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php [hunk 13]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+use Sugarcrm\Sugarcrm\custom\gradebook_fake\RecordManager;
+
+class StudentGradebookJob implements RunnableSchedulerJob {
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php]: Thanks for the detailed review! I've made the style changes you made and opened an issue to track adding a style checker to our build process.

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist (hunk 10 above)
Comment (lschaefer-sugarcrm) [reply package/src/custom/tests/School/unit-php/phpunit.xml.dist]: I removed it and the tests still ran.  I copied it out of the regular Sugar phpunit.xml.dist, so it's odd they have it there as well.

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist (hunk 11 above)
Comment (lschaefer-sugarcrm) [reply package/src/custom/tests/School/unit-php/phpunit.xml.dist]: Great catch!  I've set it (as a first start) to just get code coverage of the custom directory.  It's tricky as the custom directory gets some files when Sugar is installed that are not related to our Professor M MLP.  I'm concerned that if we get more specific with the file listings, we'll forget to update this as we add more customizations.

Here is a list of all of the files I sync from my installed Sugar directory to my git repo.  I'm thinking we don't want a list like this in phpunit.xml.dist.  What do you think?
          - /Applications/MAMP/htdocs/profm711unittestsb/tests/unit-php/custom/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/gradebook_fake/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/tests/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Schedulers/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Contacts/Ext/LogicHooks/profm.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Accounts/Accounts_Save.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Accounts/logic_hooks.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Contacts/Students_Gradebook.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/**/*professor*.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/**/*Professor*.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Accounts/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Accounts/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-accounts-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Calls/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-calls-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Calls/Ext/Vardefs/rli_link_workflow.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Campaigns/Ext/Layoutdefs/_overrideCampaign_subpanel_opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Contacts/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-contacts-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Contacts/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Documents/Ext/Layoutdefs/_overrideDocument_subpanel_opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Documents/Ext/Vardefs/rli_link_workflow.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Leads/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Meetings/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Notes/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-notes-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Notes/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Opportunities/Ext/Dependencies/opp_disable_dep.ext.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Opportunities/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Products/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-products-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Products/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Project/Ext/Layoutdefs/_overrideProject_subpanel_opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Project/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Quotes/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-quotes-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Quotes/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/RevenueLineItems/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-revenuelineitems-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/RevenueLineItems/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Tasks/Ext/clients/base/layouts/subpanels/_overridesubpanel-for-tasks-opportunities.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/Extension/modules/Tasks/Ext/Vardefs/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/icons/**/*Professor*.*
          - /Applications/MAMP/htdocs/profm711unittestsb/language/application/en_us.lang.php
          - /Applications/MAMP/htdocs/profm711unittestsb/metadata/**/*professor*.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Accounts/clients/base/views/list/list.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Accounts/clients/base/views/record/record.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Contacts/clients/base/views/list/list.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Contacts/clients/base/views/record/record.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Leads/clients/base/filters/default/default.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Leads/clients/base/layouts/convert-main/convert-main.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Leads/clients/base/views/list/list.php
          - /Applications/MAMP/htdocs/profm711unittestsb/custom/modules/Leads/clients/base/views/record/record.php
          - /Applications/MAMP/htdocs/profm711unittestsb/modules/PR_Professors/**/*
          - /Applications/MAMP/htdocs/profm711unittestsb/modules/RevenueLineItems/metadata/studio.php

File: scripts/RunProfMPHPUnitTests.sh (hunk 12 above)
Comment (lschaefer-sugarcrm) [reply scripts/RunProfMPHPUnitTests.sh]: Good catch!

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist (hunk 10 above)
Comment (glevine) [reply package/src/custom/tests/School/unit-php/phpunit.xml.dist]: I agree. I don't quite know why it was added in the first place. If you wanted to create a test suite that runs only tests that validate the logic hook handlers, then you might have a "logic hook" suite that specifies only the files and/or directories where those tests exist. But I don't think you need to worry about that yet. Running everything is fast.

File: package/src/custom/modules/Contacts/Students_Gradebook.php [hunk 14]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+require_once 'include/SugarQueue/SugarJobQueue.php';
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply package/src/custom/modules/Contacts/Students_Gradebook.php]: You're right!

File: package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+     * @var a partial mock of RecordManager
+     */
+    private $rm;
+
+    /**
+     * @var a partial mock of StudentGradebookJob
+     */
+    private $sgJob;
+
+    public function setup()
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php]: I'm surprised that the lower case version worked.  Updated.

File: package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        \SugarAutoLoader::load('../../../Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php');
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php]: I was able to change the path to 'custom/Extension/modules....'  Much easier to read!

File: package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php (hunk 2 above)
Comment (lschaefer-sugarcrm) [reply package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php]: Added.  So I know for the future...what is the benefit of namespacing unit tests?

File: package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php [hunk 17]
```diff
@@ -0,0 +1,122 @@
+<?php
+
+use Sugarcrm\SugarcrmTestsUnit\TestMockHelper;
+use Sugarcrm\Sugarcrm\Util\Uuid;
+
+/**
+ * @coversDefaultClass \StudentGradebookJob
+ */
+class StudentGradebookJobTest extends \PHPUnit_Framework_TestCase
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply package/src/custom/tests/School/unit-php/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJobTest.php]: I verified that updating to \PHPUnit\Framework\TestCase works when running the tests against Sugar 7.10 and 7.11.

**Comment by lschaefer-sugarcrm** [comment]
@glevine Thanks for the info on writing an integration test!  I filed an internal ticket as you suggested.  I also opened Issue #43 to track adding the integration test.

I think I've addressed your other suggestions.

File: package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    public function run($data)
+    {
+        if (!empty($data)) {
+            $bean = $this->getContactBean($data);
+
+            try {
+                //Call the external GradebookFake app to create a new record in it
+                $rm = $this->getRecordManager();
+                $success = $rm->createStudentRecord($bean->email1, $bean->first_name, $bean->last_name);
```
Comment (glevine) [inline package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php]: `email1` shouldn't be used anymore. `SugarEmailAddress::getPrimaryAddress()` can be used to get the string of the primary email address. You don't have to change this now, but you'll want to create an issue to remind you to do it later.

File: package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+                if ($success){
```
Comment (glevine) [inline package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php]: You should have a space before the `{`.

File: package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+                    $this->job->succeedJob();
+                    return true;
+                } else{
```
Comment (glevine) [inline package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php]: You should have a space before the `{`.

File: package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * Get the Contact (Student) bean for the given id
+     * @param $id The id for which you want to retrieve a Contact (Student) bean
+     * @return null|SugarBean
+     */
+    protected function getContactBean($id)
+    {
+        return BeanFactory::getBean('Contacts', $id);
```
Comment (glevine) [inline package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php]: `BeanFactory::getBean()` always returns a bean. If it can't find the right one by the `$id` then it just returns an empty bean of that type (Contacts). `BeanFactory::retrieveBean()` will return `null` if the record can't be found, which is what I think you want.

File: package/src/custom/modules/Contacts/Students_Gradebook.php [hunk 22]
```diff
@@ -15,30 +13,61 @@ class Students_Gradebook
...
-            $job->target = "function::AddStudentToGradebookJob";
-
-            global $current_user;
-            //set the user the job runs as
-            $job->assigned_user_id = $current_user->id;
-
-            //push into the queue to run
-            $jq = new SugarJobQueue();
-            $jobid = $jq->submitJob($job);
+        if($arguments['isUpdate']){
```
Comment (glevine) [inline package/src/custom/modules/Contacts/Students_Gradebook.php]: You should have a space after `if` and another space before `{`.

File: package/src/custom/modules/Contacts/Students_Gradebook.php [hunk 23]
```diff
@@ -15,30 +13,61 @@ class Students_Gradebook
...
+     * @param SchedulersJob $job The job to submit
+     * @return string Response from submitting the job
+     */
+    protected function scheduleJob(\SchedulersJob $job)
+    {
+        $jq = $this->getSugarJobQueue();
+        return $jq->submitJob($job);
+    }
+
+    protected function getSchedulersJob()
```
Comment (glevine) [inline package/src/custom/modules/Contacts/Students_Gradebook.php]: Missing doc block.

File: package/src/custom/modules/Contacts/Students_Gradebook.php [hunk 24]
```diff
@@ -15,30 +13,61 @@ class Students_Gradebook
...
+    {
+        return new SchedulersJob();
+    }
+
+    protected function getSugarJobQueue()
```
Comment (glevine) [inline package/src/custom/modules/Contacts/Students_Gradebook.php]: Missing doc block.

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+        <ini name="memory_limit" value="-1"/>
+    </php>
+    <testsuites>
+        <testsuite name="all">
+            <directory>./</directory>
+        </testsuite>
+    </testsuites>
+    <filter>
+        <whitelist processUncoveredFilesFromWhitelist="false">
+            <directory suffix=".php">../../</directory>
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/phpunit.xml.dist]: At this point, I think these are the only files you want to generate code coverage for. You could continue to add to `custom/gradebook_fake`. You could also make changes to `custom/Extension/modules/Schedulers/Ext/ScheduledTasks/AddStudentToGradebookJob.php` and/or `custom/modules/Contacts/Students_Gradebook.php`. So I think you should have this as your whitelist:

```
<directory suffix=".php">../../custom/gradebook_fake</directory>
<file>../../custom/Extension/modules/Schedulers/Ext/ScheduledTasks/AddStudentToGradebookJob.php</file>
<file>../../custom/modules/Contacts/Students_Gradebook.php</file>
```

That would only generate code coverage for your code. I don't think you can avoid having to maintain this list. It's just going to have to be a part of your DoD. If you add a new class to `custom/modules/Accounts`, then you'll need to add that on its own. But if you add another class to `custom/gradebook_fake`, then you won't need to add it because it'll already be captured by the `<directory>` directive.

Obviously, this is untested, so the only way to know if it works is to run it. You can either run it selectively from the command line or always run it.

To run it on the command line, you would run your tests with `/path/to/phpunit --coverage-html /path/to/output/file`. You just have to decide what type of coverage report you want to run (and the location where you want to write the file). The choices are:

```
--coverage-html /tmp/report
--coverage-clover /tmp/coverage.xml
--coverage-php /tmp/coverage.serialized
--coverage-text
> /tmp/logfile.txt
--log-junit /tmp/logfile.xml
--testdox-html /tmp/testdox.html
--testdox-text /tmp/testdox.txt
```

To run it always, it's just another addition to your config file. Again, it's just a matter of choosing the type and the path to the file for output.

```
<logging>
    <log type="coverage-html" target="/tmp/report"/>
</logging>
```

File: scripts/RunPackUnitTestsAndBuildProfMPackage.sh [hunk 26]
```diff
@@ -1,9 +1,12 @@
+# This script is designed specifically for Jenkins.  It runs the Jasmine and PHPUnit tests associated with the Pack
+# script.  It also builds the Professor M module loadable package.
+
 # If you are running Jenkins in a Docker container and mounting /var/jenkins_home to a directory on your machine's
 # local filesystem, you may need to update LOCALWORKSPACEPATH to reflect the path to the Jenkins workspace on your local
```
Comment (mmarum-sugarcrm) [inline scripts/RunPackUnitTestsAndBuildProfMPackage.sh]: What's a Jenkins workspace? Can it be any directory?

File: scripts/SetupEnvAndRunPHPUnitTests.sh [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+######################################################################
+
+if [[ -z "$1" ]] || [[ -z "$2" ]] || [[ -z "$3" ]] || [[ -z "$4" ]] || [[ -z "$5" ]] || [[ -z "$6" ]] || [[ -z "$7" ]]
+then
+    echo "Not all required command line arguments were set. Please run the script again with the required arguments:
+        1: Email address associated with your SugarCRM Developer Builds Community account
+        2: Password associated with the above account
+        3: Sugar version (Example: 7.11)
+        4: Sugar edition (Options: Ult, Ent, Pro)
+        5: GitHub username that has access to sugarcrm/unit-tests
```
Comment (mmarum-sugarcrm) [inline scripts/SetupEnvAndRunPHPUnitTests.sh]: This is a problem for me because I use 2FA. Github supports using SSH keys to authenticate. It would be good (and probably a better security practice) to have the option of using SSH keys.

File: README.md [hunk 28]
```diff
@@ -201,35 +224,125 @@ the following in [.travis.yml](.travis.yml).
...
+   1. Scope: **Global** 
+   1. Username: your SugarCRM Developer Builds Space username
+   1. Password: your SugarCRM Developer Builds Space password
+   1. ID: SUGARCRM_DEVELOPER_BUILDS_SPACE
+   1. Description: Account that has access to the SugarCRM Developer Builds Space
+1. Create a new global secret for the path to where Sugar Docker will be stored on your host machine (not the path on 
+Jenkins).
+   1. Kind: **Secret text**
+   1. Scope: **Global** 
+   1. Secret: The path to where Sugar Docker will be stored on your host machine.  For example, if your Jenkins
```
Comment (mmarum-sugarcrm) [inline README.md]: I don't really understand this piece. It's a bit tricky for me. You have to enter this path here and also in the Jenkins build script further below in a couple different paths. Is there a way to just provide a root path and then you figure out all the necessary relative paths?

File: package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php (hunk 18 above)
Comment (lschaefer-sugarcrm) [reply package/src/custom/Extension/modules/Schedulers/Ext/ScheduledTasks/StudentGradebookJob.php]: Since I was messing with the file anyway, I went ahead and fixed this.

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist (hunk 25 above)
Comment (lschaefer-sugarcrm) [reply package/src/custom/tests/School/unit-php/phpunit.xml.dist]: Thanks for the info on how to run the code coverage!  I was able to get it working.  I tweaked the settings and ended up with the following:

```
        <whitelist addUncoveredFilesFromWhitelist="false" processUncoveredFilesFromWhitelist="false">
            <directory suffix=".php">../../..</directory>
            <exclude>
                <directory suffix=".php">../..</directory>
            </exclude>
        </whitelist>
```

This sets the coverage to run on the custom directory and excludes the custom/tests directory.  The downside to this is that if someone fails to test a file at all, it won't show up in the code coverage report.  However, I feel more confident that we'll remember to check if a PR has appropriate testing than if the PR updates the phpunit.xml.dist.  We could set this file and forget about it.  What do you think?

**Comment by lschaefer-sugarcrm** [comment]
@mmarum-sugarcrm Going to start going through your comments...


```Error: No such container: my-composer```
This one is somewhat expected.  This command is removing a container if it exists from a previous build.  If the container doesn't exist (because you haven't run the build before or the previous build cleanly removed everything), you'll get this error.  I'll add some notes about it.

```
docker exec my-yarn yarn install OCI runtime exec failed: exec failed: container_linux.go:348: starting container process caused "process_linux.go:86: executing setns process caused \"exit status 21\"": unknown
```
This one is more of a mystery.  Googling didn't bring up much.  Can you update to the latest Docker and try again?

File: scripts/SetupEnvAndRunPHPUnitTests.sh [hunk 29]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env bash
+
+######################################################################
+# Variables
+######################################################################
+
+if [[ -z "$1" ]] || [[ -z "$2" ]] || [[ -z "$3" ]] || [[ -z "$4" ]] || [[ -z "$5" ]] || [[ -z "$6" ]] || [[ -z "$7" ]]
+then
+    echo "Not all required command line arguments were set. Please run the script again with the required arguments:
+        1: Email address associated with your SugarCRM Developer Builds Community account
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply scripts/SetupEnvAndRunPHPUnitTests.sh]: Agreed.  I ended up fixing this in https://github.com/sugarcrm/school/pull/44/files (not yet merged).

File: scripts/RunPackUnitTestsAndBuildProfMPackage.sh (hunk 26 above)
Comment (lschaefer-sugarcrm) [reply scripts/RunPackUnitTestsAndBuildProfMPackage.sh]: No, Jenkins creates the directory automatically for you.  Is this more clear?

If you are running Jenkins in a Docker container and mounting /var/jenkins_home to a directory on your machine's local filesystem (for example: /Users/lschaefer/jenkins), you may need to update LOCALWORKSPACEPATH to reflect the path on your local filesystem to the Jenkins workspace for your ProfessorM job (for example: "/Users/lschaefer/jenkins/workspace/ProfessorM"). The easiest way to update LOCALWORKSPACEPATH is to pass in the value as an argument when you call this script.

File: package/src/custom/tests/School/unit-php/phpunit.xml.dist [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+        <ini name="memory_limit" value="-1"/>
+    </php>
+    <testsuites>
+        <testsuite name="all">
+            <directory>./</directory>
+        </testsuite>
+    </testsuites>
+    <filter>
+        <whitelist addUncoveredFilesFromWhitelist="false" processUncoveredFilesFromWhitelist="false">
+            <directory suffix=".php">../../..</directory>
```
Comment (glevine) [inline package/src/custom/tests/School/unit-php/phpunit.xml.dist]: It's a little too magical for me. But it should work fine. It can always change later if something comes up that requires it.

File: scripts/SetupEnvAndRunPHPUnitTests.sh (hunk 27 above)
Comment (lschaefer-sugarcrm) [reply scripts/SetupEnvAndRunPHPUnitTests.sh]: Good point!  I opened #51 to track this

File: README.md (hunk 28 above)
Comment (lschaefer-sugarcrm) [reply README.md]: I opened #52 to track this

File: .travis.yml [hunk 31]
```diff
@@ -1,20 +1,69 @@
...
     before_install:
     - yarn global add grunt-cli
+  - stage: Run PHPUnit
+    sudo: required
+    language: php
+    php: '7.1'
+    services:
+    - docker
+    env:
+    - SUGAR_VERSION=7.11 SUGAR_EDITION=Ent
```
[the remark this answers is not in the document]
Comment (lschaefer-sugarcrm) [reply .travis.yml]: Notes from our chat:  We're going to continue running on ent, pro, and ult as this tests the install into each version.  I'm going to remove the running of the Sugar provided unit tests.

**Comment by lschaefer-sugarcrm** [comment]
Based on conversation I had with @mmarum-sugarcrm on Friday, I'm going to merge this PR.  I addressed the big thing he wanted me to fix (removing the Sugar unit tests from the builds) and opened new issues to track the remaining things.
